daemon/analytics.c

```diff
@@ -0,0 +1,874 @@
+}
+
+void analytics_exporters(void)
+{
+    //when no exporters are available, an empty string will be sent
+    //decide if something else is more suitable (but propably not null)
+    BUFFER *bi = buffer_create(1000);
+    analytics_exporting_connectors(bi);
+    analytics_set_data_str(&analytics_data.netdata_exporting_connectors, (char *)buffer_tostring(bi));
+    buffer_free(bi);
+}
+
+int collector_counter_callb(void *entry, void *data)
+{
+    struct array_printer *ap = (struct array_printer *)data;
+    struct collector *col = (struct collector *)entry;
+
+    BUFFER *bt = ap->both;
+
+    if (likely(ap->c)) {
+        buffer_strcat(bt, ",");
+    }
+
+    buffer_strcat(bt, "{");
+    buffer_strcat(bt, " \"plugin\": \"");
+    buffer_strcat(bt, col->plugin);
+    buffer_strcat(bt, "\", \"module\":\"");
+    buffer_strcat(bt, col->module);
+    buffer_strcat(bt, "\" }");
+
+    (ap->c)++;
+
+    return 0;
+}
+
+/*
+ * Create a JSON array of available collectors, same as in api/v1/info
+ */
+void analytics_collectors(void)
+{
+    RRDSET *st;
+    DICTIONARY *dict = dictionary_create(DICTIONARY_FLAG_SINGLE_THREADED);
+    char name[500];
+    BUFFER *bt = buffer_create(1000);
+
+    rrdset_foreach_read(st, localhost)
+    {
+        if (rrdset_is_available_for_viewers(st)) {
+            struct collector col = { .plugin = st->plugin_name ? st->plugin_name : "",
+                                     .module = st->module_name ? st->module_name : "" };
+            snprintfz(name, 499, "%s:%s", col.plugin, col.module);
+            dictionary_set(dict, name, &col, sizeof(struct collector));
+        }
+    }
+
+    struct array_printer ap;
+    ap.c = 0;
+    ap.both = bt;
+
+    dictionary_get_all(dict, collector_counter_callb, &ap);
+    dictionary_destroy(dict);
+
+    analytics_set_data(&analytics_data.netdata_collectors, (char *)buffer_tostring(ap.both));
+
+    {
+        char b[7];
+        snprintfz(b, 6, "%d", ap.c);
+        analytics_set_data(&analytics_data.netdata_collectors_count, b);
+    }
+
+    buffer_free(bt);
+}
+
+/*
+ * Run alarm-notify.sh script using the dump_methods parameter
+ * SEND_CUSTOM is always available
+ */
+void analytics_alarms_notifications(void)
+{
+    char *script;
+    script = mallocz(
+        sizeof(char) * (strlen(netdata_configured_primary_plugins_dir) + strlen("alarm-notify.sh dump_methods") + 2));
+    sprintf(script, "%s/%s", netdata_configured_primary_plugins_dir, "alarm-notify.sh");
+    if (unlikely(access(script, R_OK) != 0)) {
+        info("Alarm notify script %s not found.", script);
+        freez(script);
+        return;
+    }
+
+    strcat(script, " dump_methods");
+
+    pid_t command_pid;
+
+    info("Executing %s", script);
+
+    BUFFER *b = buffer_create(1000);
+    int cnt = 0;
+    FILE *fp = mypopen(script, &command_pid);
+    if (fp) {
+        char line[200 + 1];
+
+        while (fgets(line, 200, fp) != NULL) {
```

Comment:
We use an approach a little bit different in [health_config.c](https://github.com/netdata/netdata/blob/d7212e031da9d1f827a64b7ba9e22c30df72b8e5/health/health_config.c#L548), because the manual gives ( `man 3 fgets` gives the following information:

```
fgets() reads in at most one less than size characters from stream and stores them into the buffer pointed to by s.  Reading stops after an EOF or a newline.
```

I think we need to rework the logic here.